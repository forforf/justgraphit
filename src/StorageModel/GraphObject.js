
class GraphObject {

  static fromPlainObject(obj) {
    const ks = Object.keys(obj);
    if (ks.length > 1) {
      throw new Error(`Only single keyed object supported, but got ${ks}`)
    }
    const name = Object.keys(obj)[0];
    const data = obj[name];
    return new GraphObject(name, data);
  }

  constructor(graphName, graphData) {
    this.name = graphName;
    this.data = graphData;
  }
}

export default GraphObject