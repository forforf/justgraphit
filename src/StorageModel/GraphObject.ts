import {GraphName, JustGraphitEntry, JustGraphitStore} from "../JustGraphitTypes";


// Used to enforce that there is a single key/value pair (key = name, value = data in this case)
class GraphObject {
  name: GraphName;
  data: JustGraphitEntry[];

  // input object should look like a store, but actually only have a single graph -> data entry
  static fromPlainObject(obj: JustGraphitStore): GraphObject {
    const ks = Object.keys(obj);
    if (ks.length > 1) {
      throw new Error(`Only single keyed object supported, but got ${ks}`)
    }
    const name = Object.keys(obj)[0];
    const data = obj[name];
    return new GraphObject(name, data);
  }

  constructor(graphName: GraphName, graphData: JustGraphitEntry[]) {
    this.name = graphName;
    this.data = graphData;
  }
}

export default GraphObject;