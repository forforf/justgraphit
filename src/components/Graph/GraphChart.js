import React, {useEffect, useRef, useState} from 'react';
import GraphElement from "./GraphElement";


const GRAPH_DEFAULTS = {
  height: 200,
  width: 300
};

function GraphChart({graphObject}) {
  const [height, setHeight] = useState(GRAPH_DEFAULTS.height);
  const [width, setWidth] = useState(GRAPH_DEFAULTS.width)
  const graphRef = useRef(null);

  // Only triggered on render.
  // there might be a way to trigger graph resize when browser is resized
  // but would that be a feature or a bug?
  // Maybe we *want* the graph to stay the same size when resizing browser window. but if so need to tell user about it
  useEffect(() => {
    const current = graphRef && graphRef.current;
    const h = current && current.clientHeight;
    const w = current && current.clientWidth
    setHeight(h);
    setWidth(w)
  }, []);


  return (
    <div className="GraphChart" ref={graphRef}>
      <GraphElement graphName={graphObject.name} graphData={graphObject.data} height={height} width={width} />
    </div>
  )
}

export default GraphChart
