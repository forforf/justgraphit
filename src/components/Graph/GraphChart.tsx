import React, { useEffect, useRef, useState } from 'react';
import GraphElement from './GraphElement';
import GraphObject from '../../StorageModel/GraphObject';

const GRAPH_DEFAULTS: { height: number; width: number } = {
  height: 200,
  width: 300,
};

export type GraphChartProps = {
  graphObject: GraphObject;
};

function enforceMinimum(curVal: number, minVal: number): number {
  return curVal < minVal ? minVal : curVal
}

function GraphChart({ graphObject }: GraphChartProps): JSX.Element {
  const [height, setHeight] = useState<number>(GRAPH_DEFAULTS.height);
  const [width, setWidth] = useState<number>(GRAPH_DEFAULTS.width);
  const graphRef = useRef<HTMLDivElement>(null);

  // Only triggered on render.
  // there might be a way to trigger graph resize when browser is resized
  // but would that be a feature or a bug?
  // Maybe we *want* the graph to stay the same size when resizing browser window. but if so need to tell user about it
  useEffect(() => {
    const current = graphRef?.current;
    const h = current?.clientHeight;
    const w = current?.clientWidth;

    if (h != null && w != null) {
      // There's probably a better way to ensure that a brand-new graph is the correct size
      const hEnforced = enforceMinimum(h, GRAPH_DEFAULTS.height)
      const wEnforced = enforceMinimum(w, GRAPH_DEFAULTS.height)
      /* `!= null` checks for null and undefined */
      setHeight(hEnforced);
      setWidth(wEnforced);
    }
  }, [graphRef]);

  return (
    <div className="GraphChart" ref={graphRef}>
      <GraphElement
        graphData={graphObject.data}
        height={height}
        width={width}
      />
    </div>
  );
}

export default GraphChart;
