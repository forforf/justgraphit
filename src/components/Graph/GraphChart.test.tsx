import React from 'react';
import renderer from 'react-test-renderer';
import ReactDom from 'react-dom';
import { act } from 'react-dom/test-utils';
import GraphChart, { GraphChartProps } from './GraphChart';
import GraphObject from '../../StorageModel/GraphObject'; // let mockD3select: () => jest.Mock<unknown, any>;

let graphChartProps: GraphChartProps;
beforeEach(() => {
  graphChartProps = {
    // graphObject: jest.fn(() => ({name: 'foo', data: []}))
    graphObject: { name: 'foo', data: [] } as GraphObject,
  };
});

test('deep snapshot', () => {
  const component = renderer.create(<GraphChart {...graphChartProps} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('useEffect', () => {
  const mount = document.createElement('div');
  act(() => {
    ReactDom.render(<GraphChart {...graphChartProps} />, mount);
  });
  expect(mount).toMatchSnapshot();
});
