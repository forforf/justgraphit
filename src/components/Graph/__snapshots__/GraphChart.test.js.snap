// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep snapshot 1`] = `
<div
  className="GraphChart"
>
  <div>
    This is an EmptyGraph
  </div>
</div>
`;

exports[`useEffect 1`] = `
<div>
  <div
    class="GraphChart"
  >
    <div>
      This is an EmptyGraph
    </div>
  </div>
</div>
`;
